rfc/system/3749-sane-overridable-defaults.md

```diff
@@ -0,0 +1,208 @@
+# Sane overridable defaults
+## Background
+Flyte comes with a rich set of overridable defaults ([matchable resources](https://docs.flyte.org/en/latest/deployment/configuration/general.html#deployment-configuration-general)) from everything to default task resource requests, project resource quota, kubernetes labels and annotations for executions and default service accounts.
+
+These can be configured to have default levels across a Flyte installation, on a per-domain basis, on a per-project basis and event at a per-workflow basis.
+
+Managing these requires setting the defaults at a level of specifity using a variety of [message types](https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/admin/matchable_resource.proto#L14,L38). Furthermore it's difficult to holistically reason about what overrides exist at say, the project level. Updating these overrides is error-prone because there is no existing merge logic and requires understanding the data model to even know what specific matchable resource to specifically update.
+
+Importantly, the current mechanism for applying application overrides is driven imperatively using flytectl and hard to reason about across revisions or manage using [Infrastructure as Code](https://learn.microsoft.com/en-us/devops/deliver/what-is-infrastructure-as-code).
+
+## Proposal
+We should be able to make it simple to view, edit and update defaults for domains, projects and workflows. It should be difficult to unintentionally overwrite changes and easy to understand what is currently applied.
+
+## Overview
+
+### Proposed User Experience
+It should be simple to view overridable defaults
+
+```
+flytectl get overrides -o overrides.yaml
+...
+
+flytectl get overrides -d development -p flytesnacks -o overrides.yaml
+...
+
+```
+For
+* global/per-domain defaults (immutable application settings)
+* project defaults
+* per-workflow defaults
+
+
+#### Proposed document structure
+A single call to get all overrides should return the collated document for all overrides set at a level (e.g. per project)
+
+```
+$ flytectl get overrides -p flytesnacks -d development --workflow MyExampleWorkflow
+
+apiVersion: 1  # Version of this document structure
```

Comment:
Not a CRD I assume despite the similarity, right?

Reply:
No, borrowing heavily on k8s for structuring this object but not proposing moving to a CRD just yet in this RFC :)

Reply:
Makes sense :)